### PR TITLE
Allow generating URLs with extra parameters

### PR DESCRIPTION
**Pull request by emesik**
This simple extension allows generating signed URLs with custom parameters, e.g. changing content disposition or cache control for individual links.

### CODE REVIEW

**Comment by michielvermeir** [comment]
Oops, I didn't see your PR. I _just_ submitted another PR, because I have a similar need. Just one question though: How will you pass these extra params be passed from the Django model? The `FileField.url` property calls the storage provider's `url()` method behind the scenes, no?

**Comment by emesik** [comment]
I have a simple wrapper function in my project:

```
def s3_url_as_attachment(fieldfile):
    return fieldfile.storage.url(
        fieldfile.name,
        extra_params={'ResponseContentDisposition': 'attachment'})
```

and use it by calling `s3_url_as_attachment(obj.filefield)`

I saw your solution, and while it leaves no place for fine-grained control, I like the idea that it applies the configuration from `settings` instantly without any need to modify metadata on S3.

I wonder what @etianen thinks about all that.

**Comment by etianen** [comment]
Hello everyone :)

Let's use this PR to discuss the two approaches. Both seem to work similarly, by adding parameters to the presigned URL directly, rather than setting metadata on the underlying object.

I'm a little worried that this will be a source of user confusion, since it's common to use `AWS_S3_PUBLIC_URL` to serve non-presigned assets from a public read-only bucket or CDN, which gives better caching characteristics.

It might be safer to set the custom headers on the object directly, since then it'll work using both presigned URLs, and static public URLs. Since there's a pretty efficient `s3_sync_meta` management command to apply these settings to existing objects, it's not a huge burden to the developer.

In any case, to avoid settings proliferation, maybe a callable setting called `AWS_S3_EXTRA_PARAMS` could be used, which can be used to set content disposation or any other headers that might be required?

**Comment by emesik** [comment]
I haven't noticed the `s3_sync_meta` command and it might be a solution to my particular problem. I'm still unsure how it works, though. The inner `s3_connection.copy_content` suggests it might transfer entire object data, i.e. using a lot of bandwidth and incurring significant fees.

However, I was looking for a solution that would be easy to start with but possible to apply in complex scenarios. I believe that approach _all objects in a bucket share the same metadata_ wouldn't apply in many situations and users would like to have custom URLs for different objects within a bucket, or even multiple parametrized URLs to the same object. Therefore, there should be a way to generate custom URLs that:

1. Applies extra parameters no matter if the default URL for the bucket is signed or public (if possible).
2. When called without extra params it would apply default settings, overriding metadata the object has attached within S3.
3. Accept all possible extra options to allow generating any potential URL that S3 could handle.

What do you think about it?

BTW, I'm wondering how you launch tests in this project. I haven't figured it out yet.

And sorry for raising three different questions in one comment ;)

**Comment by etianen** [comment]
I don't think that copy_content() transfers any data when the source and
destination are the same bucket. I think it's just an API call sent to S3,
so the data never leaves the S3 bucket.

Allowing the url() method to take extra params for customizing URL
generation seems reasonable, since then you can use it to generate all
sorts of custom metadata in the URL. The docs would need updating to make
it clear that it won't work if you use the AWS_S3_PUBLIC_URL setting.

The travis CI file shows how to launch tests:

https://github.com/etianen/django-s3-storage/blob/master/.travis.yml

Look in the env/global section for the environmental variables that need to
be present to run the tests.

**Comment by emesik** [comment]
Updated the docs and added tests. Please check if it's OK.

File: django_s3_storage/storage.py [hunk 1]
```diff
@@ -130,7 +130,7 @@ class S3Storage(Storage):
     """
 
     default_auth_settings = {
-        "AWS_REGION": "us-east-1",
+        "AWS_REGION": "eu-west-1",
```
Comment (etianen) [inline django_s3_storage/storage.py]: I don't think you meant to change this part...

```suggestion
        "AWS_REGION": "eu-east-1",
```

File: django_s3_storage/storage.py [hunk 2]
```diff
@@ -396,14 +396,19 @@ def size(self, name):
         except KeyError:
             return meta["ContentLength"]
 
-    def url(self, name):
+    def url(self, name, extra_params=None):
         # Use a public URL, if specified.
         if self.settings.AWS_S3_PUBLIC_URL:
+            if extra_params:
+                raise ValueError("Use of extra_params to generate custom URLs is not allowed "
+                        "with AWS_S3_PUBLIC_URL")
```
Comment (etianen) [inline django_s3_storage/storage.py]: Nice touch!

File: README.rst [hunk 3]
```diff
@@ -183,6 +183,30 @@ To make media files public, and enable aggressive caching, make the following ch
...
+.. code:: python
+
+    url = storage.url("foo/bar.pdf", extra_params={"ResponseContentDisposition": "attachment"})
+
+Another example is a link to a specific version of the file (within a bucket that has versioning
+enabled):
+
+.. code:: python
+
+    url = storage.url("foo/bar.pdf", extra_params={"FRy3fTduRtqHsRAoNp0REzPJj_WunDfl"})
```
Comment (etianen) [inline README.rst]: This is a set, not a dict, did you mean:

```suggestion
    url = storage.url("foo/bar.pdf", extra_params={"VersionId": "FRy3fTduRtqHsRAoNp0REzPJj_WunDfl"})
```

File: django_s3_storage/storage.py (hunk 1 above)
Comment (emesik) [reply django_s3_storage/storage.py]: Absolutely. This got in by accident.

**Comment by emesik** [comment]
Corrected. Thanks for quick review!

**Comment by etianen** [comment]
Thanks! I took out the `testCustomUrlVersionId` test, since my test bucket doesn't support versioning and it wasn't testing and code paths not covered by `testCustomUrlContentDisposition`.

**Comment by emesik** [comment]
Well, you may enable versioning in the bucket by a single click in Properties. But okay :)

**Comment by etianen** [comment]
Does it cost more though? Does it actually let me delete stuff, or wil it
just grow larger over the course of my tests as object versions accumulate?

I can't be bothered right now to go looking through AWS documentation,
basically.

**Comment by emesik** [comment]
The old versions are kept after `delete()`. This way I've salvaged a couple of files lost to a bug. I don't know how to delete all permanently via `boto`; would have to look it up in the docs too.

**Comment by etianen** [comment]
Released as 0.12.5

**Pull request merged**